### PR TITLE
Add `aws_redshift_scheduled_action` resource

### LINKED ISSUE #12745: Add support for Redshift scheduled actions
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It looks like there is currently no support for AWS Redshift scheduled actions such as pause and resume (those actions and schedules are different to those of snapshots) — at least I could not find any references to them in docs nor in issue tracker here.

### New or Affected Resource(s)

* aws_redschift_scheduled_action

### Potential Terraform Configuration

```hcl
resource "aws_redschift_scheduled_action" "this" {
  name = "nightly-pause"
  description = "Pause redshift cluster for the night"  # optional
  active = true  # optional

  # possibly use some more user-friendly time format here
  start_time = "2020-04-10T00:00:00+00:00"  # optional
  end_time = "2020-05-11T00:00:00+00:00"  # optional
  schedule = "cron(00 21 * * ? *)"   # time is UTC
  iam_role = aws_iam_role.scheduler.arn 

  target_action {
    action = "PauseCluster"
    cluster_identifier = aws_redshift_cluster.this.id
  }
}
```

### References

* https://docs.aws.amazon.com/redshift/latest/APIReference/API_ScheduledAction.html

### PR DESCRIPTION
**Pull request by waneal**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #12745 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
New Resource: aws_redshift_scheduled_action
```

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSRedshiftScheduledAction'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRedshiftScheduledAction -timeout 120m
=== RUN   TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
=== PAUSE TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
=== RUN   TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
=== PAUSE TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
=== CONT  TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
=== CONT  TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
--- PASS: TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions (51.01s)
--- PASS: TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions (51.24s)
--- PASS: TestAccAWSRedshiftScheduledActionResizeCluster_basic (75.16s)
--- PASS: TestAccAWSRedshiftScheduledActionResumeCluster_basic (75.98s)
--- PASS: TestAccAWSRedshiftScheduledActionPauseCluster_basic (76.85s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       79.310s
```

### CODE REVIEW

File: aws/resource_aws_redshift_scheduled_action.go
Comment (waneal) [inline aws/resource_aws_redshift_scheduled_action.go:145]: Redshift scheduled action does not have arn.

File: aws/resource_aws_redshift_scheduled_action.go
Comment (waneal) [inline aws/resource_aws_redshift_scheduled_action.go:183]: If each of StartTime and EndTime are not set, "0001-01-01T00:00:00Z" was returned. So when it was returned, start_time and end_time are not set.

Comment (bflad) [reply aws/resource_aws_redshift_scheduled_action.go:183]: You may be able to check against `scheduledAction.StartTime != nil` and `scheduledAction.EndTime != nil` instead 👍

Comment (waneal) [reply aws/resource_aws_redshift_scheduled_action.go]: @bflad Thank you for telling, I fixed it!

**Comment by spaszek** [comment]
So, any chance of merging it soon?

**Comment by waneal** [comment]
@bflad Do I have to do more things? Or wait for review?

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by waneal** [comment]
I will update by end of September!

**Comment by waneal** [comment]
```
> make testacc TEST=./aws TESTARGS='-run=TestAccAWSRedshiftScheduledAction'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRedshiftScheduledAction -timeout 120m
=== RUN   TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
=== PAUSE TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
=== RUN   TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== PAUSE TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== RUN   TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
=== PAUSE TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
=== CONT  TestAccAWSRedshiftScheduledActionPauseCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionResizeCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions
=== CONT  TestAccAWSRedshiftScheduledActionResumeCluster_basic
=== CONT  TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions
--- PASS: TestAccAWSRedshiftScheduledActionResizeCluster_WithOptions (70.56s)
--- PASS: TestAccAWSRedshiftScheduledActionPauseCluster_WithOptions (72.91s)
--- PASS: TestAccAWSRedshiftScheduledActionResumeCluster_basic (107.14s)
--- PASS: TestAccAWSRedshiftScheduledActionResizeCluster_basic (107.36s)
--- PASS: TestAccAWSRedshiftScheduledActionPauseCluster_basic (107.87s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       109.269s
```

**Comment by waneal** [comment]
Can someone review this please?

**Comment by tedivm** [comment]
I'd love to be able to use this or somethign similar. AWS has [updated the functionality](https://aws.amazon.com/about-aws/whats-new/2020/10/amazon-redshift-supports-scheduling-sql-queries-by-integrating-with-amazon-eventbridge/) around this as well.

**Comment by mohsen0** [comment]
Hi Guys, any update on this PR?

**Comment by waneal** [comment]
1 year has passed since this pull request got ready, please someone review

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTARGS='-run=TestAccAWSRedshiftScheduledAction_'                                                      
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRedshiftScheduledAction_ -timeout 180m
=== RUN   TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== RUN   TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
=== PAUSE TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
=== RUN   TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== RUN   TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== RUN   TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== PAUSE TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== RUN   TestAccAWSRedshiftScheduledAction_disappears
=== PAUSE TestAccAWSRedshiftScheduledAction_disappears
=== CONT  TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== CONT  TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== CONT  TestAccAWSRedshiftScheduledAction_disappears
=== CONT  TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== CONT  TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== CONT  TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
--- PASS: TestAccAWSRedshiftScheduledAction_disappears (28.22s)
--- PASS: TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions (31.07s)
--- PASS: TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions (31.21s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicResumeCluster (45.87s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicPauseCluster (45.88s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicResizeCluster (45.89s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	49.841s
```

##### GovCloud

```console
% make testacc TESTARGS='-run=TestAccAWSRedshiftScheduledAction_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRedshiftScheduledAction_ -timeout 180m
=== RUN   TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== RUN   TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
=== PAUSE TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
=== RUN   TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== RUN   TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== PAUSE TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== RUN   TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== PAUSE TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== RUN   TestAccAWSRedshiftScheduledAction_disappears
=== PAUSE TestAccAWSRedshiftScheduledAction_disappears
=== CONT  TestAccAWSRedshiftScheduledAction_basicPauseCluster
=== CONT  TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions
=== CONT  TestAccAWSRedshiftScheduledAction_basicResizeCluster
=== CONT  TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions
=== CONT  TestAccAWSRedshiftScheduledAction_basicResumeCluster
=== CONT  TestAccAWSRedshiftScheduledAction_disappears
--- PASS: TestAccAWSRedshiftScheduledAction_disappears (32.61s)
--- PASS: TestAccAWSRedshiftScheduledAction_PauseClusterWithOptions (35.95s)
--- PASS: TestAccAWSRedshiftScheduledAction_ResizeClusterWithOptions (36.04s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicPauseCluster (51.05s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicResizeCluster (51.12s)
--- PASS: TestAccAWSRedshiftScheduledAction_basicResumeCluster (53.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	56.301s
```

**Comment by waneal** [comment]
@ewbankkit Thank you for fixing!

**Pull request merged**
